cranelift/isle/isle/src/trie.rs

```diff
@@ -196,28 +207,78 @@ impl TrieNode {
 
         // If we are empty, turn into a decision node.
         if self.is_empty() {
-            *self = TrieNode::Decision { edges: vec![] };
+            *self = TrieNode::Decision {
+                edges: vec![],
+                last_prio: None,
+                cur_prio: None,
+            };
         }
 
         // We must be a decision node.
-        let edges = match self {
-            &mut TrieNode::Decision { ref mut edges } => edges,
+        let (edges, last_prio, cur_prio) = match self {
+            &mut TrieNode::Decision {
+                ref mut edges,
+                ref mut last_prio,
+                ref mut cur_prio,
+            } => (edges, last_prio, cur_prio),
             _ => panic!("insert on leaf node!"),
         };
 
+        // If we are inserting at a lower prio than in `cur_prio`,
+        // `cur_prio` moves to `last_prio` (and controls our minimum
+        // insertion point) and we initialize `cur_prio` with our
+        // current priority and new max insertion index (of 0).
+        let first_in_cur = if cur_prio.is_none() || prio < cur_prio.unwrap().0 {
+            *last_prio = *cur_prio;
+            *cur_prio = Some((prio, 0));
+            true
+        } else {
+            false
+        };
+        let cur_prio = cur_prio.as_mut().unwrap();
+
+        // Determine the minimum edge index under which we can insert
+        // while respecting priorities.
+        let start = if let Some((last_prio, last_index)) = *last_prio {
+            assert!(last_prio > prio);
```

Comment:
Is the priority value in `last_prio` tracked only for this assert?

Reply:
It was, yeah, and I guess as a side-effect of using the same type for `cur_prio` and `last_prio` (it's actually needed for `cur_prio`). This `let start =` expression got much simpler in the refactor though so it's no longer here.